# $\alpha$=0.1

In [4]:
import numpy as np
import torch
import torchvision.transforms as transforms               # include image preprocess tools
from torchvision.datasets import CIFAR10        # for loading images from Pytorch CIFAR
from torch.utils.data import DataLoader
import detectors
import timm
from src.saps import saps_test
from src.temperature_scaling import ModelWithTemperature

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# load fine-tuned model
model = timm.create_model("resnet50_cifar10", pretrained=True)
model = model.to(device)

# reprocess the images from CIFAR
data_transform = transforms.Compose([
    transforms.ToTensor(),          # transfer to tensor
    transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])  # normalize
])
# load images from CIFAR10
dataset = CIFAR10(root="../../data", train=False, download=True, transform=data_transform)

# temperature scaling
temp_scal_loader = DataLoader(dataset, batch_size=32, shuffle=True)
model = ModelWithTemperature(model, temperature=5.0).to(device)
model.set_temperature(temp_scal_loader)

model.eval()

saps_test(model, dataset, device, num_runs=10, alpha=0.1, lambda_=2.0)

Files already downloaded and verified
Before temperature - NLL: 0.354, ECE: 0.046
Optimal temperature: 4.908
After temperature - NLL: 0.827, ECE: 0.420
SAPS Classification, Start!

Running experiment 1/10...
t_cal = 0.5995162010192873
Average Prediction Set Size After APS in runs 1: 1.0292
Average Coverage Rate in runs 1: 0.907

Running experiment 2/10...
t_cal = 0.5961227893829347
Average Prediction Set Size After APS in runs 2: 1.0256
Average Coverage Rate in runs 2: 0.8998

Running experiment 3/10...
t_cal = 0.5932057261466983
Average Prediction Set Size After APS in runs 3: 1.0202
Average Coverage Rate in runs 3: 0.8984

Running experiment 4/10...
t_cal = 0.5880305886268617
Average Prediction Set Size After APS in runs 4: 1.0148
Average Coverage Rate in runs 4: 0.8934

Running experiment 5/10...
t_cal = 0.5916524291038514
Average Prediction Set Size After APS in runs 5: 1.0218
Average Coverage Rate in runs 5: 0.9014

Running experiment 6/10...
t_cal = 0.5987095177173615
Average Pre

# $\alpha$=0.05

In [5]:
saps_test(model, dataset, device, num_runs=10, alpha=0.05, lambda_=2.0)

Files already downloaded and verified
Before temperature - NLL: 0.354, ECE: 0.046
Optimal temperature: 4.908
After temperature - NLL: 0.827, ECE: 0.420
SAPS Classification, Start!

Running experiment 1/10...
t_cal = 1.4840792357921606
Average Prediction Set Size After APS in runs 1: 1.4954
Average Coverage Rate in runs 1: 0.951

Running experiment 2/10...
t_cal = 1.2669203996658327
Average Prediction Set Size After APS in runs 2: 1.3724
Average Coverage Rate in runs 2: 0.944

Running experiment 3/10...
t_cal = 1.5182092189788823
Average Prediction Set Size After APS in runs 3: 1.505
Average Coverage Rate in runs 3: 0.9548

Running experiment 4/10...
t_cal = 1.238810002803804
Average Prediction Set Size After APS in runs 4: 1.3702
Average Coverage Rate in runs 4: 0.9442

Running experiment 5/10...
t_cal = 1.5079073369503024
Average Prediction Set Size After APS in runs 5: 1.5026
Average Coverage Rate in runs 5: 0.9526

Running experiment 6/10...
t_cal = 1.4597425162792241
Average Predic

# Result
$\alpha$=0.1
- Final Average **Prediction Set Size：1.02 / 10**
- Final Average **Coverage: 90.09%**  

$\alpha$=0.05
- Final Average **Prediction Set Size：1.46 / 10**
- Final Average **Coverage: 95.01%**